File: ProjectArchive/SourceCode/Java/DataStructures/Find.java
package DataStructures;

import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.TreeSet;
import java.util.concurrent.TimeUnit;

public class Find {

    @State(Scope.Benchmark)
    public static class States {

        @Param({"10", "100", "1000"})
        public int noOfElements;

        public int succ = 0;
        public int fail = 0;

        public int[] arr;
        public ArrayList<Integer> arrList;
        public LinkedList<Integer> linkedList;
        public HashSet<Integer> hashSet;
        public TreeSet<Integer> treeSet;

        @Setup(Level.Invocation)
        public void setup() {
            arrList = new ArrayList<>();
            linkedList = new LinkedList<>();
            hashSet = new HashSet<>();
            treeSet = new TreeSet<>();
            arr = new int[noOfElements];
            succ = noOfElements / 2;
            fail = noOfElements + 1;
            for(int i = 0; i < noOfElements; i++){
                arr[i] = i;
                arrList.add(i);
                linkedList.add(i);
                hashSet.add(i);
                treeSet.add(i);
            }
        }
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void succArrayFind(States state, Blackhole blackhole){
        for(int i = 0; i < state.noOfElements; i++){
            if(state.arr[i] == state.succ){
                break;
            }
        }
        blackhole.consume(state.arr);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void failArrayFind(States state, Blackhole blackhole){
        for(int i = 0; i < state.noOfElements; i++){
            if(state.arr[i] == state.fail){
                break;
            }
        }
        blackhole.consume(state.arr);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void succArrayListFind(States state, Blackhole blackhole){
        state.arrList.contains(state.succ);
        blackhole.consume(state.arrList);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void failArrayListFind(States state, Blackhole blackhole){
        state.arrList.contains(state.fail);
        blackhole.consume(state.arrList);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void succLinkedListFind(States state, Blackhole blackhole){
        state.linkedList.contains(state.succ);
        blackhole.consume(state.linkedList);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void failLinkedListFind(States state, Blackhole blackhole){
        state.linkedList.contains(state.fail);
        blackhole.consume(state.linkedList);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void succHashSetFind(States state, Blackhole blackhole){
        state.hashSet.contains(state.succ);
        blackhole.consume(state.hashSet);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void failHashSetFind(States state, Blackhole blackhole){
        state.hashSet.contains(state.fail);
        blackhole.consume(state.hashSet);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void succTreeSetFind(States state, Blackhole blackhole){
        state.treeSet.contains(state.succ);
        blackhole.consume(state.treeSet);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(value = 1, warmups = 1)
    public void failTreeSetFind(States state, Blackhole blackhole){
        state.treeSet.contains(state.fail);
        blackhole.consume(state.treeSet);
    }
}

File: ProjectArchive/SourceCode/C++/DataStructures/Find.cpp
#include <benchmark/benchmark.h>
#include <queue>
#include <list>
#include <stack>
#include <vector>
#include <deque>

static void succArrayFind(benchmark::State &state){
    for(auto _ : state){

        state.PauseTiming();
        int arr[state.range(0)];
        for(int i = 0; i < state.range(0); i++){
            arr[i] = i;
        }   
        int succ = state.range(0) / 2;
        state.ResumeTiming();

        for(int i = 0; i < state.range(0); i++){
            if(arr[i] == succ){
                break;
            }
        }
    }
    benchmark::ClobberMemory();
}
BENCHMARK(succArrayFind)->Arg(10)->Arg(100)->Arg(1000);

static void failArrayFind(benchmark::State &state){
    for(auto _ : state){

        state.PauseTiming();
        int arr[state.range(0)];
        for(int i = 0; i < state.range(0); i++){
            arr[i] = i;
        }   
        int fail = state.range(0) + 2;
        state.ResumeTiming();

        for(int i = 0; i < state.range(0); i++){
            if(arr[i] == fail){
                break;
            }
        }
    }
    benchmark::ClobberMemory();
}
BENCHMARK(failArrayFind)->Arg(10)->Arg(100)->Arg(1000);


static void succVectorFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::vector<int> vec;
        for(int i = 0; i < state.range(0); i++){
            vec.push_back(i);
        }
        int succ = state.range(0) / 2;
        state.ResumeTiming();

        std::find(vec.begin(), vec.end(), succ);
    }
    benchmark::ClobberMemory();
}
BENCHMARK(succVectorFind)->Arg(10)->Arg(100)->Arg(1000);

static void failVectorFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::vector<int> vec;
        for(int i = 0; i < state.range(0); i++){
            vec.push_back(i);
        }
        int fail = state.range(0) + 2;
        state.ResumeTiming();

        std::find(vec.begin(), vec.end(), fail);
    }
    benchmark::ClobberMemory();
}
BENCHMARK(failVectorFind)->Arg(10)->Arg(100)->Arg(1000);

static void succListFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::list<int> lst;  
        for(int i = 0; i < state.range(0); i++){
            lst.push_back(i);
        }  
        int succ = state.range(0) / 2;
        benchmark::ClobberMemory();
        state.ResumeTiming();

        std::find(lst.begin(), lst.end(), succ);
    }
    benchmark::ClobberMemory();
}
BENCHMARK(succListFind)->Arg(10)->Arg(100)->Arg(1000);

static void failListFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::list<int> lst;  
        for(int i = 0; i < state.range(0); i++){
            lst.push_back(i);
        }  
        int fail = state.range(0) + 2;
        benchmark::ClobberMemory();
        state.ResumeTiming();

        std::find(lst.begin(), lst.end(), fail);
    }
    benchmark::ClobberMemory();
}
BENCHMARK(failListFind)->Arg(10)->Arg(100)->Arg(1000);

static void succQueueFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::deque<int> q;  
        for(int i = 0; i < state.range(0); i++){
            q.push_back(i);
        }
        int succ = state.range(0) / 2;
        benchmark::ClobberMemory();
        state.ResumeTiming();

        std::find(q.begin(), q.end(), succ);        
    }
    benchmark::ClobberMemory();
}
BENCHMARK(succQueueFind)->Arg(10)->Arg(100)->Arg(1000);

static void failQueueFind(benchmark::State &state){
    for(auto _ : state){
        state.PauseTiming();
        std::deque<int> q;  
        for(int i = 0; i < state.range(0); i++){
            q.push_back(i);
        }
        int fail = state.range(0) + 2;
        benchmark::ClobberMemory();
        state.ResumeTiming();

        std::find(q.begin(), q.end(), fail); 
    }
    benchmark::ClobberMemory();
}
BENCHMARK(failQueueFind)->Arg(10)->Arg(100)->Arg(1000);

BENCHMARK_MAIN();